docs-v2/design_proposals/buildx-transparent-support.md

```diff
@@ -0,0 +1,284 @@
+to have different cache destinations (production vs development cache, with different permissions).
+
+Multi-platform combined image builds are directly supported by buildx, including improved caching of common layers and emulation / cross-compilation.
+This could simplify pipelines and provide faster builds of complex codebases.
+
+References:
+
+* https://www.docker.com/blog/image-rebase-and-improved-remote-cache-support-in-new-buildkit/
+* https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/
+
+## Proposal
+
+This proposal aims to improve Skaffold's build process by adding transparent support for docker buildx.
+The primary goal is to enable users to leverage the advanced features of buildx, such as remote buildkit builders and improved caching, without significant changes to their existing Skaffold configurations.
+
+New Global Configs:
+
+* `buildx-builder`: Enables automatic detection of buildx and multiple builder support.
+* `cache-tag`: Allows overriding the default cache tagging strategy, useful for managing caches across different branches or environments.
+
+Skaffold Schema changes:
+
+* `cacheTo` to specify custom cache destinations, adding `--cache-to` support to the Docker CLI build process when using buildx.
+
+Build Process Enhancements:
+
+* The build process now intelligently detects and utilizes buildx if available and configured.
+* Reduced dependency on the local Docker daemon when using buildx, enhancing security and flexibility.
+
+Backward Compatibility:
+
+All changes are designed to be backward compatible.
+If buildx is not detected or used, Skaffold will fall back to the traditional Docker builder.
+
+This cache behavior is intended to provide transparent user experience, with similar functionality compared to traditional local docker builds,
+without additional boilerplate or different configuration / patches for a CI workflow.
+
+## Design approach
+
+`docker buildx build` can be configured to execute against a [remote buildkit instance](https://docs.docker.com/build/builders/drivers/remote/).
+No docker daemon is necessary for this to work, but also that is supported by default using the [docker driver](https://docs.docker.com/build/builders/drivers/docker/).
+Additionally, [docker container driver](https://docs.docker.com/build/builders/drivers/docker-container)
+or [kubernetes driver](https://docs.docker.com/build/builders/drivers/kubernetes/) are available too for advanced use cases.
+
+Since `docker build` and `docker buildx build` essentially share the same options, no major modifications are needed to Skaffold for backward compatibility.
+
+This proposal implements the logic to detect if buildx is the default builder (looking for an alias in the docker config).
+To [set buildx as the default builder](https://github.com/docker/buildx?tab=readme-ov-file#set-buildx-as-the-default-builder), the command `docker buildx install` should be used.
```

Comment:
Not a must-have by any means, but it'd be nice to have an explicit way at the skaffold.yml level, like `useBuildkit: true`, to specify that buildx *should* be used, and therefore error out if its not available.

Reply:
A more actionable error would be great if buildx is not installed properly (that is included in the design doc for further phases), but I explicitly would like to avoid a configuration in skaffold.yaml because:
* users that doesn't have buildx installed would still be able to build
* no modification of skaffold.yaml is needed to switch between CI and local envs

This way, buildx support is transparent and the user can decide to use it or not, just setting the config parameter with `skaffold config set buildx-builder ...`

Yes, some advanced features would not work (like exporting the cache), but this would not be critical for local development (and wont work either if proper registry push credentials are not given).

Reply:
That thinking makes sense to me, I do like your idea of providing graceful "fallback". That said, iIn the cases of very big projects, if cache-from in particular isn't working, I'd personally prefer my "app developers" to get an error that they _really_ need to configure buildx, at the very least a noisy warning. Many of them will silently not configure buildx, and will later complain that builds take 30 minutes, when if cache-from was working, it'd be like 5 minutes.

`skaffold config` setting the builder system-wide rather than per-project feels like it has small issues in some cases: for example, in my case I will have one project where I need multi-platform, and another where I don't. But I agree that hardcoding the builder names in skaffold.yaml would be much worse.